### ISSUE TITLE
Training YOLOv8 without writing images to disk

### ISSUE THREAD

**Issue by pabsan-0** [ultralytics/ultralytics#22281]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi, I have a dataset in a format other than images, from which I want to generate pictures without writing them to disk, in order to train a YOLOv8 model. 

I have followed similar issues, mainly [#10174](https://github.com/ultralytics/ultralytics/issues/10174), and as instructed I have implemented a custom dataset with proper `__len__` and `__getitem__` methods. I can now properly generate the training images from my private data. However, I can't figure out the part about using such dataset for training, since that part is entirely skipped in the issue above with only pseudocode indications.

Any input on that regard?

So far, I have tried using a custom detection trainer, however ultralytics still requests a valid YAML file pointing to images. If I could use any custom format of YAML file to take advantage of its fields that'd be nice, however bypassing the yaml file requirement would be good enough for now. 

This is the training code I'm playing with.
```
import torch
from torch.utils.data import Dataset

class VirtualImageDataset(Dataset):
    def __init__(self, file_list=None):
        # Placeholder storage
        self.entries = [0, 1, 2]  # dummy length

    def __len__(self):
        return len(self.entries)

    def __getitem__(self, idx):
        # Hardcoded dummy image and bounding boxes
        fake_image = torch.ones((3, 128, 128), dtype=torch.float32)  # RGB image
        fake_boxes = torch.tensor([[0, 0.5, 0.5, 0.2, 0.2]])  # [cls, cx, cy, w, h]

        return {
            "image": fake_image,      # [3,H,W]
            "bboxes": fake_boxes,    # [N,5]
        }


class MyCustomTrainer(DetectionTrainer):
    def build_dataset(self, img_path, mode="train", batch=None):
        return VirtualImageDataset()

    def get_dataloader(self, dataset_path, batch_size, rank=0, mode="train"):
        dataset = self.build_dataset(dataset_path, mode=mode, batch=batch_size)

        return DataLoader(
            dataset,
            batch_size=batch_size,
            shuffle=(mode == "train"),
            num_workers=self.args.workers,
            pin_memory=True,
            drop_last=True,
        )
      
if __name__ == "__main__":
    model = YOLO("yolov8n.pt")
    results = model.train(trainer=MyCustomTrainer, data="hello.yaml", epochs=2)
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @pabsan-0, thanks for opening the issue and for your interest in Ultralytics 🚀! This is an automated response to help get things moving quickly—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including:
-Your ultralytics package version and environment details (OS, Python, PyTorch, CUDA)
-Full code snippet showing how you initialize the model and trainer
-Exact error messages or stack traces you’re seeing
-A small reproducible snippet of your custom Dataset output format (e.g., a single sample) and any custom Trainer overrides
-Training logs and a brief description of the expected vs. actual behavior
And please verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Thanks for the detailed context—yes, you can train directly from a PyTorch `Dataset` without writing images or using a YAML; pass a minimal `data` dict so the trainer knows class names and return the batch keys YOLO expects (`img`, `cls`, `bboxes`, `batch_idx`):

```python
import torch
from torch.utils.data import Dataset, DataLoader
from ultralytics import YOLO
from ultralytics.models.yolo.detect.train import DetectionTrainer

class VirtualImageDataset(Dataset):
    def __len__(self):
        return 3

    def __getitem__(self, _):
        # img as uint8 [C,H,W] (0..255); trainer will .float()/255
        img = torch.full((3, 128, 128), 127, dtype=torch.uint8)
        # boxes as [class, x, y, w, h] with xywh normalized to 0..1
        boxes = torch.tensor([[0, 0.5, 0.5, 0.2, 0.2]], dtype=torch.float32)
        return {
            "img": img,                 # [3,H,W], uint8 or float 0..255
            "cls": boxes[:, :1],        # (N,1)
            "bboxes": boxes[:, 1:],     # (N,4) normalized xywh
        }

def collate_fn(batch):
    imgs = torch.stack([b["img"] for b in batch], 0)
    cls = torch.cat([b["cls"] for b in batch], 0)
    bboxes = torch.cat([b["bboxes"] for b in batch], 0)
    batch_idx = torch.cat(
        [torch.full((b["cls"].shape[0], 1), i, dtype=torch.int64) for i, b in enumerate(batch)], 0
    )
    return {"img": imgs, "cls": cls, "bboxes": bboxes, "batch_idx": batch_idx}

class MyCustomTrainer(DetectionTrainer):
    def build_dataset(self, img_path, mode="train", batch=None):
        return VirtualImageDataset()

    def get_dataloader(self, dataset_path, batch_size, rank=0, mode="train"):
        ds = self.build_dataset(dataset_path, mode=mode, batch=batch_size)
        return DataLoader(ds, batch_size=batch_size, shuffle=(mode == "train"),
                          num_workers=0, collate_fn=collate_fn, drop_last=True)

model = YOLO("yolov8n.pt")
model.train(trainer=MyCustomTrainer, data={"names": ["obj"], "nc": 1}, epochs=2, imgsz=128)
```

The same approach applies to val/test by checking `mode` in your trainer; for the standard call signature and dataset formats see the training examples in the [Object Detection Datasets Overview](https://docs.ultralytics.com/datasets/detect/).

**Comment by pabsan-0** [comment]
Hi @glenn-jocher , thanks for your reply. I'm having multiple issues from here:

- Passing a dict like that does not work. I bypassed this by creating a dummy file, adding some extra fields that the library requested.
- The library requires additional `Dataset` attributes: at least `labels` and `loss_items`.

These issues do not align with the given snippet. Is this really the way to go or are we missing something?

If I may, @Y-T-G once showed me a neat tip involving custom dataset classes for an unrelated issue, maybe his input could be valuable?

**Comment by Y-T-G** [comment]
@pabsan-0 You should subclass `YOLODataset` instead of the torch `Dataset` class

**Comment by jrusbo** [comment]
@Y-T-G The problem with subclassing from `YOLODataset` is that it inherits from `BaseDataset` which depends heavily on the image files rather than the images themselves.

**Comment by Y-T-G** [comment]
Ultralytics expects a lot of those attributes to be present, that's why it is the base. So you will eventually have to account for them. You can override the attributes and methods as needed.
